kaldb/src/main/java/com/slack/kaldb/bulkIngestApi/BulkIngestApi.java

```diff
@@ -0,0 +1,103 @@
+
+/**
+ * This class is responsible for defining the http endpoint behavior for the bulk ingest. It is
+ * expected to handle appropriate rate limiting, error handling, and submit the parsed messages to
+ * Kafka for ingestion.
+ */
+public class BulkIngestApi {
+  private static final Logger LOG = LoggerFactory.getLogger(BulkIngestApi.class);
+  private final BulkIngestKafkaProducer bulkIngestKafkaProducer;
+  private final DatasetRateLimitingService datasetRateLimitingService;
+  private final MeterRegistry meterRegistry;
+  private final Counter incomingByteTotal;
+  private final Timer bulkIngestTimer;
+  private final String BULK_INGEST_INCOMING_BYTE_TOTAL = "kaldb_preprocessor_incoming_byte";
+  private final String BULK_INGEST_TIMER = "kaldb_preprocessor_bulk_ingest";
+
+  public BulkIngestApi(
+      BulkIngestKafkaProducer bulkIngestKafkaProducer,
+      DatasetRateLimitingService datasetRateLimitingService,
+      MeterRegistry meterRegistry) {
+
+    this.bulkIngestKafkaProducer = bulkIngestKafkaProducer;
+    this.datasetRateLimitingService = datasetRateLimitingService;
+    this.meterRegistry = meterRegistry;
+    this.incomingByteTotal = meterRegistry.counter(BULK_INGEST_INCOMING_BYTE_TOTAL);
+    this.bulkIngestTimer = meterRegistry.timer(BULK_INGEST_TIMER);
+  }
+
+  @Post("/_bulk")
+  public HttpResponse addDocument(String bulkRequest) {
+    // 1. Kaldb does not support the concept of "updates". It's always an add.
+    // 2. The "index" is used as the span name
+    CompletableFuture<HttpResponse> future = new CompletableFuture<>();
+    Timer.Sample sample = Timer.start(meterRegistry);
+    future.thenRun(() -> sample.stop(bulkIngestTimer));
+
+    try {
+      byte[] bulkRequestBytes = bulkRequest.getBytes(StandardCharsets.UTF_8);
+      incomingByteTotal.increment(bulkRequestBytes.length);
+      Map<String, List<Trace.Span>> docs = BulkApiRequestParser.parseRequest(bulkRequestBytes);
+
+      // todo - our rate limiter doesn't have a way to acquire permits across multiple
+      // datasets
+      // so today as a limitation we reject any request that has documents against
+      // multiple indexes
+      // We think most indexing requests will be against 1 index
+      if (docs.keySet().size() > 1) {
+        BulkIngestResponse response =
+            new BulkIngestResponse(0, 0, "request must contain only 1 unique index");
+        future.complete(HttpResponse.ofJson(INTERNAL_SERVER_ERROR, response));
+        return HttpResponse.of(future);
+      }
+
+      for (Map.Entry<String, List<Trace.Span>> indexDocs : docs.entrySet()) {
+        final String index = indexDocs.getKey();
+        if (!datasetRateLimitingService.tryAcquire(index, indexDocs.getValue())) {
+          BulkIngestResponse response = new BulkIngestResponse(0, 0, "rate limit exceeded");
+          future.complete(HttpResponse.ofJson(TOO_MANY_REQUESTS, response));
+          return HttpResponse.of(future);
+        }
+      }
+
+      Thread.ofVirtual()
```

Comment:
Discussed this offline, could we for now just leave a comment saying long term we might want to remove the `Thread.ofVirtual` from here and just have the blocking task executor be be backed by virtual threads.